receiver/collectdreceiver/collectd.go

```diff
@@ -0,0 +1,292 @@
+// limitations under the License.
+
+package collectdreceiver
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/timestamp"
+)
+
+const (
+	collectDMetricDerive   = "derive"
+	collectDMetricGauge    = "gauge"
+	collectDMetricCounter  = "counter"
+	collectDMetricAbsolute = "absolute"
+)
+
+type collectDRecord struct {
+	Dsnames        []*string              `json:"dsnames"`
+	Dstypes        []*string              `json:"dstypes"`
+	Host           *string                `json:"host"`
+	Interval       *float64               `json:"interval"`
+	Plugin         *string                `json:"plugin"`
+	PluginInstance *string                `json:"plugin_instance"`
+	Time           *float64               `json:"time"`
+	TypeS          *string                `json:"type"`
+	TypeInstance   *string                `json:"type_instance"`
+	Values         []*json.Number         `json:"values"`
+	Message        *string                `json:"message"`
+	Meta           map[string]interface{} `json:"meta"`
+	Severity       *string                `json:"severity"`
+}
+
+func (r *collectDRecord) isEvent() bool {
+	return r.Time != nil && r.Severity != nil && r.Message != nil
+}
+
+func (r *collectDRecord) protoTime() *timestamp.Timestamp {
+	if r.Time == nil {
+		return nil
+	}
+	ts := time.Unix(0, int64(float64(time.Second)**r.Time))
+	tsp, err := ptypes.TimestampProto(ts)
+	if err != nil {
+		return nil
+	}
+	return tsp
+}
+
+func (r *collectDRecord) extractAndAppendMetrics(metrics []*metricspb.Metric, defaultLabels map[string]string) ([]*metricspb.Metric, error) {
+	// Ignore if record is an event instead of data point
+	if r.isEvent() {
+		recordEventsReceived()
+		return metrics, nil
+
+	}
+
+	recordMetricsReceived()
+	labels := make(map[string]string, len(defaultLabels))
+	for k, v := range defaultLabels {
+		labels[k] = v
+	}
+
+	for i := range r.Dsnames {
+		if i < len(r.Dstypes) && i < len(r.Values) && r.Values[i] != nil {
+			dsType, dsName, val := r.Dstypes[i], r.Dsnames[i], r.Values[i]
+			metricName, usedDsName := r.getReasonableMetricName(i, labels)
+
+			addIfNotNullOrEmpty(labels, "plugin", true, r.Plugin)
+			parseAndAddLabels(labels, r.PluginInstance, r.Host)
+			addIfNotNullOrEmpty(labels, "dsname", !usedDsName, dsName)
+
+			metric, _ := r.newMetric(metricName, dsType, val, labels)
+			metrics = append(metrics, metric)
+
+		}
+	}
+	return metrics, nil
+}
+
+func (r *collectDRecord) newMetric(name string, dsType *string, val *json.Number, labels map[string]string) (*metricspb.Metric, error) {
+	metric := &metricspb.Metric{}
+	point, isDouble, err := r.newPoint(val)
+	if err != nil {
+		return metric, err
+	}
+
+	lKeys, lValues := labelKeysAndValues(labels)
+	metric.MetricDescriptor = &metricspb.MetricDescriptor{
+		Name:      name,
+		Type:      r.metricType(dsType, isDouble),
```

Comment:
Comment: this bothers me on the proto: the point has the value but the metric descriptor too... so the format itself allow inconsistencies... am I missing something?

Reply:
It does. In the sense that descriptor can say the metric is let's say commulative double but each point can be then be either int or double.

Reply:
No, not improved. Otel metrics still has a type enum in the decsriptor and then a set of fields to store data points. It is possible to misuse it the same way we have here.

Reply:
Is it worth opening an issue on spec repo? Or there are reasons for that?